### PR TITLE
share MemOptVarInfos of external variables into cinn_launch subgraph

### PR DESCRIPTION
**Pull request by CtfGo**
### PR types

Others
### PR changes

Others
### Describe

1.  add a graph pass named `share_varinfo_into_cinn_pass`: it finds all cinn_launch subgraphs on the main graph and share MemOptVarInfos of their external variables from the outer global graph to the subgraph.
2. add an data member named `parent_holder` on the `MemOptVarInfo` class, which is used in a cinn_launch subgraph and points to the MemOptVarInfo handle of the same variable from the main graph.
3. upgrade the condition of `EagerDeletionOpHandle` erasing variable: when it appears in a subgraph, it should check the MemOptVarInfo from the main graph

### CODE REVIEW

File: paddle/fluid/framework/details/eager_deletion_op_handle.cc [hunk 1]
```diff
@@ -107,6 +107,19 @@ void EagerDeletionOpHandle::CallOnce() {
 
 std::string EagerDeletionOpHandle::Name() const { return "eager_deletion"; }
 
+static bool CanBeErased(ir::MemOptVarInfo *var_info) {
+  if (var_info->IsSkippedAllMemoryOptimization() ||
+      !var_info->DecreaseRefCnt()) {
+    return false;
+  }
+  auto parent_holder = var_info->ParentHolder();
```
Comment (zhhsplendid) [inline paddle/fluid/framework/details/eager_deletion_op_handle.cc]: Personal opinion: `auto` is not good when the code readers haven't look at ParentHolder. Suggest to replace it with real type.

Reference: https://google.github.io/styleguide/cppguide.html#Type_deduction

File: paddle/fluid/framework/ir/memory_optimize_pass/share_mem_opt_info_to_subgraph_pass.cc [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      }
+    }
+  }
+  return nullptr;
+}
+
+static void ShareVarInfoToCinnLaunch(
+    const MemOptVarInfoMapList& var_infos,
+    details::ComputationOpHandle* cinn_launch_op) {
+  auto* followed_eager_deletion_op =
```
Comment (zhhsplendid) [inline paddle/fluid/framework/ir/memory_optimize_pass/share_mem_opt_info_to_subgraph_pass.cc]: Will there just be 1 eager_deletion_op or not only 1?

File: paddle/fluid/framework/details/eager_deletion_op_handle.cc [hunk 3]
```diff
@@ -107,6 +107,19 @@ void EagerDeletionOpHandle::CallOnce() {
...
+  // if parent_holder exists, it should meet deletion condition too.
```
Comment (thisjiang) [inline paddle/fluid/framework/details/eager_deletion_op_handle.cc:116]: 这样写也改变了已有的逻辑吧？在没开启CINN的情况下确定不会有影响么？要不这儿也加个
```c++
#ifdef PADDLE_WITH_CINN
```

File: paddle/fluid/framework/ir/memory_optimize_pass/share_mem_opt_info_to_subgraph_pass.cc [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    if (parent_it != parent_var_infos.end()) {
+      VLOG(4) << "MemOptVarInfo of var[" << var_name << "] set parent holder";
+      cur_it->second->SetParentHolder(parent_it->second);
+    }
+  }
+}
+
+// This pass will be applied on both the main graph and all subgraphs,
+// and it distinguishs them according to whether the graph has the
+// kMemOptVarInfoFromMainGraph attribute or not.
```
Comment (zhhsplendid) [inline paddle/fluid/framework/ir/memory_optimize_pass/share_mem_opt_info_to_subgraph_pass.cc]: If the pass is only used in CINN, should we use another name? (Distinguish from cinn subgraph and non-cinn sub_graph)

File: paddle/fluid/framework/ir/memory_optimize_pass/share_mem_opt_info_to_subgraph_pass.cc [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+// On a subgraph, it iterates each variable that will be deleted by a
+// eager_deletion op, and take the MemOptVarInfo from the main graph
+// if such one found.
+class ShareMemOptInfoToSubGraphPass : public ir::Pass {
+ protected:
+  void ApplyImpl(ir::Graph* graph) const override {
+    auto all_ops = ir::FilterByNodeWrapper<details::OpHandleBase>(*graph);
+    const auto& var_infos = Get<MemOptVarInfoMapList>(kMemOptVarInfoMapList);
+
+    // the main graph
```
Comment (zhhsplendid) [inline paddle/fluid/framework/ir/memory_optimize_pass/share_mem_opt_info_to_subgraph_pass.cc]: Is that possible a non-cinn subgraph can call a cinn subgraph?

File: paddle/fluid/framework/details/eager_deletion_op_handle.cc (hunk 3 above)
Comment (CtfGo) [reply paddle/fluid/framework/details/eager_deletion_op_handle.cc:116]: 不改变，现有逻辑下 parent_holder 为空，所以会 return ture。不过用PADDLE_WITH_CINN看起来是个更好选择，已修改。

File: paddle/fluid/framework/ir/memory_optimize_pass/share_mem_opt_info_to_subgraph_pass.cc (hunk 2 above)
Comment (CtfGo) [reply paddle/fluid/framework/ir/memory_optimize_pass/share_mem_opt_info_to_subgraph_pass.cc]: every cinn_launch_op only have one at most

File: paddle/fluid/framework/ir/memory_optimize_pass/share_mem_opt_info_to_subgraph_pass.cc (hunk 4 above)
Comment (CtfGo) [reply paddle/fluid/framework/ir/memory_optimize_pass/share_mem_opt_info_to_subgraph_pass.cc]: that sounds more reasonable and i change to  `share_varinfo_into_cinn_pass` instead.

File: paddle/fluid/framework/details/eager_deletion_op_handle.cc (hunk 1 above)
Comment (CtfGo) [reply paddle/fluid/framework/details/eager_deletion_op_handle.cc]: good advice, i have changed it

File: paddle/fluid/framework/ir/memory_optimize_pass/share_mem_opt_info_to_subgraph_pass.cc (hunk 5 above)
Comment (CtfGo) [reply paddle/fluid/framework/ir/memory_optimize_pass/share_mem_opt_info_to_subgraph_pass.cc]: no, i update name of this pass and some comments to indicate clearly this pass will use with CINN

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Pull request merged**